### ISSUE TITLE
Linear and Angular speed and acceleration limits (open loop)

### ISSUE THREAD

**Issue by efernandez** [pal-robotics-forks/ros_controllers#12]
IMHO, we should put linear and angular velocity and acceleration limits in the diff_drive_controller directly. This controller receives velocity commands as input, so it looks like the place to limit them.

It seems like the safer strategy, because it's the lowest level node taking velocity commands. If we rely on another node at a higher level controlling that limits, we still leave a back-door to run without limits.

Finally, at ros_control low-level we have velocity and acceleration limits, but that's for the wheels themselves. It doesn't take into account if the robot is moving forward or rotating.

I'd like to know if you agree with setting those limits at the diff_drive_controller, before implementing it.

**Comment by bmagyar** [comment]
Warm welcome to the idea!

**Comment by adolfo-rt** [comment]
I think it's OK to enforce limits in task space, because it allows you to do things you can't do at the joint level. I would make this optional, though.

**Comment by po1** [comment]
+1 for the optional part. I'd like to point out that `0.0` might not be a good choice to mean _no limit_, as one might want to restrain the robot from rotating, for example. If that can be considered legitimate behavior, then we will need a boolean to enable/disable these limits.

**Comment by efernandez** [comment]
Mimicking the paremeters that we already have for the joint_limits, we can have some motion limits as well, with exactly the same names:

<pre>
  linear:
    has_velocity_limits: true
    max_velocity: 1.0
    has_acceleration_limits: true
    max_acceleration: 0.8
  angular:
    has_velocity_limits: true
    max_velocity: 1.7
    has_acceleration_limits: true
    max_acceleration: 1.5
</pre>

This will go with the parameters that we already have in mobile_base_controller.yaml (reem_controller_configuration/config).

The only problem is that currently we can specified the max velocity when going backwards (e.g. -0.5). Maybe, we could use split the 'linear' YAML key into 'forward' and 'backwards' motion. The result could be something like:

<pre>
  forward:
    has_velocity_limits: true
    max_velocity: 1.0
    has_acceleration_limits: true
    max_acceleration: 0.8
  backwards:
    has_velocity_limits: true
    max_velocity: 0.5
    has_acceleration_limits: true
    max_acceleration: 0.4
  angular:
    has_velocity_limits: true
    max_velocity: 1.7
    has_acceleration_limits: true
    max_acceleration: 1.5
</pre>

Additionally, all these could be inside a 'motion_limits' (or similar) key.

IMHO, the most clear solution will be the last one, with the parameters inside a 'motion_limits' key. This could be a file in the <robot>_description config folder, as the joint_limits.yaml, so we'll have a motion_limits.yaml one.

What do you think?
@po1 @bmagyar @adolfo-rt

**Comment by po1** [comment]
This raises the question of _how does the diff-drive controller know about what direction is forward?_
So far, I believe that the choice was arbitrarily made. You have to look in the code to know which motor is assumed to be on the left hand side. And the directions are arbitrarily decided, too.

Is there a convention somewhere?

**Comment by efernandez** [comment]
IMHO, the direction is forward as long as the cmd_vel linear x component is greater than 0; and backwards when it's lower than 0.
In the diff-drive controller this velocity is translated into the angular velocity applied to each wheel. For going forward, it will be positive angular velocity for each wheel; for going backwards, it will be negative.
Maybe I'm missing something, but I don't see a problem there.

**Comment by lucamarchionni** [comment]
I agree with Enrique. 
This means that the assumption should be ensured by the lower software layer (ros_control).

**Comment by po1** [comment]
Enrique: yeah, I forgot that the diff-drive controller had the knowledge of left and right wheel, and that this mapping to actual motors had to be done somewhere else. Never mind.

**Comment by adolfo-rt** [comment]
We could also use a limits specification that could be shared by other base controllers that are not diff-drives (like holonomic bases or even quadcopters). How about being explicit about x, y and z?, it's more general (good) but more verbose (not so good) as well. This example combines the structure of the `geometry_msgs::Twist` message with that of a joint limits specification:

``` yaml
# Whenever a min_* is unspecified, default to -max_*
linear:
  x:
    has_velocity_limits: true
    max_velocity: 1.0 # m/s
    min_velocity: -0.5 #m/s
    has_acceleration_limits: true
    max_acceleration: 0.8 # m/s^2
    min_acceleration: -0.4 # m/s^2
angular:
  z:
    has_velocity_limits: true
    max_velocity: 1.7 # rad/s
    has_acceleration_limits: true
    max_acceleration: 1.5 # rad/s^2
```

**Comment by po1** [comment]
I like what Adolfo suggests better. The _forward_ and _backward_ terms, even if they are obvious here, are making an assumption about the structure of the robot. And the distinction of these two is unnecessary.

**Comment by adolfo-rt** [comment]
...and we can reuse this description once we kickstart the project for following race camels with a UAV

**Comment by efernandez** [comment]
I also like Adolfo suggestion. I'll start implementing it.

Just note there's a typo (max_acceleration is repeated for linear.x, the second should be min_acceleration).

**Comment by adolfo-rt** [comment]
> Just note there's a typo (max_acceleration is repeated for linear.x, the second should be min_acceleration).

Fixed, thanks.

**Comment by bmagyar** [comment]
Had time to get back into the conversation. I like the +/- way a lot more than the forward/backwards approach. +1

**Comment by po1** [comment]
Work in progress in PR #16

**Comment by po1** [comment]
About the control policy for the acceleration limits: the velocity is _estimated_ by the odometry publisher, and uses a  rolling window average of 10 samples to filter it. It means a _LOT_ of delay.

I strongly suggest switching to an open-loop policy. I gave it a quick try and on simulation it works beautifully. I'll open a PR for that soon.

**Comment by po1** [comment]
Will close the issue when the open-loop limiters are implemented

**Comment by po1** [comment]
Open loop limiter implemented in pull request #23

**Comment by po1** [comment]
All PRs merged. Closing issue.
